src/coreComponents/fileIO/Outputs/TimeHistoryOutput.cpp

```diff
@@ -43,69 +45,64 @@ TimeHistoryOutput::TimeHistoryOutput( string const & name,
-      m_io[ioCount]->updateCollectingCount( md.getDims( )[0] );
-      return m_io[ioCount]->getBufferHead( );
-    } );
-    m_io.back()->init( !freshInit );
-    ++ioCount;
-  }
+      HistoryMetadata metadata = hc.getMetaData( domain, collectorIdx );
 
-  localIndex metaCollectorCount = collector.getNumMetaCollectors( );
-  for( localIndex metaIdx = 0; metaIdx < metaCollectorCount; ++metaIdx )
-  {
-    HistoryCollection & metaCollector = collector.getMetaCollector( metaIdx );
-    for( localIndex ii = 0; ii < metaCollector.getCollectionCount( ); ++ii )
-    {
-      HistoryMetadata metaMetadata = metaCollector.getMetadata( domain, ii );
-      metaMetadata.setName( collector.getTargetName() + " " + metaMetadata.getName( ) );
-      m_io.emplace_back( std::make_unique< HDFHistIO >( outputFile, metaMetadata, m_recordCount ) );
-      metaCollector.registerBufferCall( ii, [this, ii, ioCount, &domain, &metaCollector] ()
+      // TODO Deal with this in PackCollection::buildMetaDataCollectors?
+      if( !prefix.empty() )
+      { metadata.setName( prefix + metadata.getName() ); }
+
+      m_io.emplace_back( std::make_unique< HDFHistIO >( outputFile, metadata ) );
+//      m_io.emplace_back( std::make_unique< TestingBuffer >( metadata.getName(), metadata.getDims() ) );
+      hc.registerBufferProvider( collectorIdx, [this, idx = m_io.size() - 1]( localIndex count )
       {
-        metaCollector.updateSetsIndices( domain );
-        HistoryMetadata md = metaCollector.getMetadata( domain, ii );
-        m_io[ ioCount ]->updateCollectingCount( md.getDims()[0] );
-        return m_io[ ioCount ]->getBufferHead( );
+        m_io[idx]->updateCollectingCount( count );
+        return m_io[idx]->getBufferHead();
       } );
-      m_io.back()->init( !freshInit );
-      ++ioCount;
+      m_io.back()->init( m_hasRecordStarted );
     }
+  };
+
+  // FIXME Why stop (pseudo) recursion at one single level?
+  registerBufferCalls( collector );
 
+  for( localIndex metaIdx = 0; metaIdx < collector.numMetaDataCollectors(); ++metaIdx )
+  {
+    registerBufferCalls( collector.getMetaDataCollector( metaIdx ), collector.getTargetName() + " " );
   }
 
   // do the time output last so its at the end of the m_io list, since writes are parallel we need
   //  the rest of the collectors to share position in the list across the world comm
 
   // rank == 0 does time output for the collector
-  if( MpiWrapper::commRank( ) == 0 )
+  if( MpiWrapper::commRank() == 0 )
   {
-    HistoryMetadata timeMetadata = collector.getTimeMetadata( );
-    m_io.emplace_back( std::make_unique< HDFHistIO >( outputFile, timeMetadata, m_recordCount, 1, 2, MPI_COMM_SELF ) );
-    collector.registerTimeBufferCall( [this]() { return m_io.back()->getBufferHead( ); } );
-    m_io.back()->init( !freshInit );
+    HistoryMetadata timeMetadata = collector.getTimeMetaData();
+    m_io.emplace_back( std::make_unique< HDFHistIO >( outputFile, timeMetadata, MPI_COMM_SELF ) );
```

Comment:
On restart it should be nonzero, which was used to set the write head in the file-writer (TimeHistHDF.cpp), which was required to append to previously collected time-history files on a restart. Is that accommodated in a different way now? 

We could just assume any preexisting time history data wrote to the correct row index and just start appending from there, but IIRC there was some amount of hdf dataspace dimensionality checking to ensure a preexisting file with preexisting data was structured correctly (don't 100% remember if that included a restart row check).

Reply:
Thx @wrtobin for the comments.

My main concern about this was to trying to remove `BufferedHistoryIO::getBufferCount`.
While `m_recordCount` surely makes sense in the in an _HDF5 context_ (and I did  not remove it from the class), I'm not sure that it should be exposed outside.

For instance, my other little impl `TestingBuffer` still has to expose `getBufferCount`.
Would there be a way to let this into `HDF5` (where it belongs I think)?

Reply:
I absolutely agree that it really only makes sense at present in an HDF5 context. 

Unfortunately we have to add something to the DataRepository to be able to restart write/read, which this value necessarily requires.

I was hoping to keep the file writers (right now just HDF5, but maybe eventually additional formats as well) "stateless" across restarts, to allow the higher-level HistoryOutput to manage the state required for the restart. 

Adding some data from the file writers to the DataRepository (in this case just the 'row' count) to the data repo is really probably the correct solution long-term. In the initial development I just erred toward a "stateless" approach versus encapsulating the term in the appropriate context.

Reply:
Could the information be recomputed instead?

Reply:
We can assume that the file has written to the correct row and append from there, since on successful shutdown the file is truncated to the correct number of rows. This doesn't allow us to check on restart that the old data is where we expect to be, but will successfully append data regardless, and that data will be correct. The most likely failure mode is that a file isn't successfully truncated for some reason and we wind up with zero rows in both the 'time' itself and the actual collected data.

I can't think of a straightforward mechanism to calculate the number of 'rows' collected on a previous run other than relying on the files themselves (particularly since I think collection cadence, etc. could change on restart) over simply counting them, which is what `m_recordCount` was for.

Reply:
We could also add an additional value to the file itself telling us the expected number of rows. Not sure that would provide anything over just assuming the file truncated correctly though.

Reply:
I've rolled back because it seems more complicated than what my PR is supposed to be.
I've commented that something may be done about it nevertheless.
I think this is ready for final review. I've split some files and renamed some classes to reach more consistent naming. But I changed nothing from the code.